src/ERC20PaymasterV06.sol

```diff
@@ -0,0 +1,240 @@
+         _requireFromEntryPoint();
+        return _validatePaymasterUserOp(userOp, userOpHash, maxCost);
+    }
+
+    /// @inheritdoc IPaymaster
+    function postOp(
+        PostOpMode mode,
+        bytes calldata context,
+        uint256 actualGasCost
+    ) external override {
+        _requireFromEntryPoint();
+        _postOp(mode, context, actualGasCost);
+    }
+
+    /*´:°•.°+.*•´.*:˚.°*.˚•´.°:°•.°•.*•´.*:˚.°*.˚•´.°:°•.°+.*•´.*:*/
+    /*                ERC-4337 PAYMASTER FUNCTIONS                */
+    /*.•°:°.´+˚.*°.˚:*.´•*.+°.•°:´*.´•*.•°.•°:°.´:•˚°.*°.˚:*.´+°.•*/
+    function _validatePaymasterUserOp(UserOperation calldata userOp, bytes32 userOpHash, uint256 maxCost)
+        internal
+        returns (bytes memory context, uint256 validationResult)
+    {
+        (uint8 mode, bytes calldata paymasterConfig) = _parsePaymasterAndData(userOp.paymasterAndData);
+
+        // valid modes are 0, 1, 2, 3
+        if (mode >= 4) {
+            revert PaymasterDataModeInvalid();
+        }
+
+        uint192 tokenPrice = getPrice();
+        uint256 tokenAmount;
+        {
+            uint256 maxFeePerGas = userOp.maxFeePerGas;
+            if (mode == 0 || mode == 1) {
+                tokenAmount = (maxCost + (refundPostOpCost) * maxFeePerGas) * priceMarkup * tokenPrice
+                    / (1e18 * PRICE_DENOMINATOR);
+            } else {
+                tokenAmount = (maxCost + (refundPostOpCostWithGuarantor) * maxFeePerGas) * priceMarkup * tokenPrice
+                    / (1e18 * PRICE_DENOMINATOR);
+            }
+        }
+
+        if (mode == 0) {
+            SafeTransferLib.safeTransferFrom(address(token), userOp.sender, address(this), tokenAmount);
+            context = abi.encodePacked(tokenAmount, tokenPrice, userOp.sender, userOpHash);
+            validationResult = 0;
+        } else if (mode == 1) {
+            if (paymasterConfig.length != 32) {
+                revert PaymasterDataLengthInvalid();
+            }
+            if (uint256(bytes32(paymasterConfig[0:32])) == 0) {
+                revert TokenLimitZero();
+            }
+            if (tokenAmount > uint256(bytes32(paymasterConfig[0:32]))) {
+                revert TokenAmountTooHigh();
+            }
+            SafeTransferLib.safeTransferFrom(address(token), userOp.sender, address(this), tokenAmount);
+            context = abi.encodePacked(tokenAmount, tokenPrice, userOp.sender, userOpHash);
+            validationResult = 0;
+        } else if (mode == 2) {
+            if (paymasterConfig.length < 32) {
+                revert PaymasterDataLengthInvalid();
+            }
+
+            address guarantor = address(bytes20(paymasterConfig[0:20]));
+
+            bool signatureValid = SignatureChecker.isValidSignatureNow(
+                guarantor,
+                getHash(userOp, uint48(bytes6(paymasterConfig[20:26])), uint48(bytes6(paymasterConfig[26:32])), 0),
```

Comment:
I would parse out the paymaserConfig bytes into local variables instead of slicing to improve readability.